File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int rows = grid.size();
        int cols = grid[0].size();
        int count = 0;
        vector<vector<bool>> visited(rows, vector<bool>(cols, false));
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (visited[i][j] || grid[i][j] == '0') 
                    continue;
                count += 1;
                bfs(grid, visited, i, j);
            }
        }
        return count;
    }
    void bfs(vector<vector<char>>& grid , vector<vector<bool>>& visited, int i, int j) {
        int n_row = grid.size();
        int n_col = grid[0].size();
        queue<pair<int, int>> q;
        q.push({i,j});
        while (!q.empty()) {
            auto target = q.front();
            int row = target.first;
            int col = target.second;
            if (row - 1 >= 0) {
                if (!visited[row-1][col] && grid[row - 1][col] == '1') {
                    visited[row - 1][col] = true;
                    q.push({row - 1, col});
                }
            }
            if (row + 1 < n_row) {
                if (!visited[row + 1][col] && grid[row + 1][col] == '1') {
                    visited[row + 1][col] = true;
                    q.push({row + 1, col});
                }
            }
            if (col - 1 >= 0) {
                if (!visited[row][col - 1] && grid[row][col - 1] == '1') {
                    visited[row][col - 1] = true;
                    q.push({row, col - 1});
                }
            }
            if (col + 1 < n_col) {
                if (!visited[row][col + 1] && grid[row][col + 1] == '1') {
                    visited[row][col + 1] = true;
                    q.push({row, col + 1});
                }
            }
            q.pop();
        }
    }
};

File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        # loop through grid, for each land do a bfs/dfs search and mark land as visited

        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        n_row = len(grid)
        n_col = len(grid[0])
        count = 0

        visited = [[False] * n_col for _ in range(n_row)]

        queue = deque()

        for i in range(n_row):
            for j in range(n_col):
                if grid[i][j] == "1" and not visited[i][j]:
                    queue.append((i, j))
                    visited[i][j] = True
                    while queue:
                        x, y = queue.popleft()
                        # print(f"Curr_grid: {i}, {j} --- exploring: [{x},{y}]")
                        for direction in directions:
                            new_row = x + direction[0]
                            new_col = y + direction[1]
                            if 0 <= new_row < n_row and 0 <= new_col < n_col and grid[new_row][new_col] == "1" and not visited[new_row][new_col]:
                                queue.append((new_row, new_col))
                                visited[new_row][new_col] = True
                    count += 1

        return count